### PR TITLE
Fix too strict interleaved/unique layout check

### PR DESCRIPTION
**Pull request by stiepan**
## Description

closes 

User facing utilities that write into mdspans (mdspan copy, fill_bytes) used to rely on `tensor_query/__has_interleaved_stride_order` check to determine if all indicies in the dst tensor are mapped to unique memory offsets. If not, they errored out.

However that check was too strict, blocking valid cases that arise from slicing a tensor with |step| > 1 (see examples below).
After discussion with @fbusato and @mhoemmen we decided to get rid of the custom `__has_interleaved_stride_order` check and instead rely on the `is_unique` property. This PR:

* Implements proper is_unique check for layout_stride_relaxed (it used to always return false - it is allowed by the standard to return false is giving accurate answer turns out to be not feasible). The check is conservative: it returns false for all non-unique layouts as well as some tricky unique cases. However, it correctly indentifies as unique all layouts that result from permuting and slicing of layout_right and layout_left, which should cover all typical layout use cases.
  * note, for device code, the check still just returns false - I don't see immediate need for it on device and I wasn't sure if it is worth to run sorting of the strides in the device code for that.
* Adjusts the uniqueness precondition (only checked if assertions are on) in layout_stride. Mapping in layout_stride is supposed to be unique by definition. However, the implemented [precondition from cpp standard draft](https://eel.is/c++draft/views.multidim#mdspan.layout.stride.cons-4.3) is too strict. This PR adjusts that check to make sure that permuting valid mdspan or creating slices with submdspan cannot create layout that doesn't meet the precondition.
* Adjusts layout_stride_relaxed properties tests (to include various positive and negative edge cases for uniqueness) and layout_stride strides test (to ensure constructing a valid sliced layout is possible with the adjusted precondition).
* Removes `__has_interleaved_stride_order` and replaces its usages with `dst.is_unique()`.

Using Python snippets for brivety:
```
# start with shape (2, 5) and take every second element along the right extent
a = np.zeros((2, 5), dtype=np.uint8)[:, ::2]
a.shape == (2, 3)
a.strides == (5, 2) 

# the old check would return true, because 2 * 3 > 5 (smallest strides * extent > next stride)
# incorrectly concluding the layout is overlapping
``` 

The new check, is still a conservative approximation (never misses interleaved layout but may flag too much as interleaved),
but returns less false postives and should handle all layouts that are a result of permuting, slicing, reshaping or broadcasting dense layout.

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by Jacobfaib** [comment]
@mhoemmen could you take a look at this?

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh [hunk 1]
```diff
@@ -32,6 +32,8 @@
 #  include <cuda/experimental/__copy_bytes/mdspan_to_raw_tensor.cuh>
 #  include <cuda/experimental/__copy_bytes/types.cuh>
 
+#  include <iostream>
```
Comment (fbusato) [inline cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: please remove this header

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh [hunk 2]
```diff
@@ -145,26 +148,50 @@ template <typename _Tp, typename _Extents, typename _LayoutPolicy, typename _Acc
...
+    // restricted to i - 1 dims with smallest strides
+    // max_offset = sum((e_j - 1) * s_j for j in range(i) if s_j > 0)
+    // min_offset = sum((e_j - 1) * s_j for j in range(i) if s_j < 0)
+    // max_dist = max_offset - min_offset = sum((e_j - 1) * |s_j| for j in range(i))
+    __stride_t __max_dist = 0;
+    for (__rank_t __i = 0; __i < __rank; ++__i)
     {
-      const auto __extent = static_cast<__stride_t>(__extents[__i]);
-      if (__extent * cudax::__abs_integer(__strides[__i]) > cudax::__abs_integer(__strides[__i + 1]))
+      const __stride_t extent = __extents[__i];
```
Comment (fbusato) [inline cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]:
```suggestion
      const auto __extent = static_cast<__stride_t>(__extents[__i]);
```

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh [hunk 3]
```diff
@@ -145,26 +148,50 @@ template <typename _Tp, typename _Extents, typename _LayoutPolicy, typename _Acc
...
+      if (extent != 1)
       {
-        return true;
+        const auto __abs_stride = cudax::__abs_integer(__strides[__i]);
+        if (__abs_stride <= __max_dist)
+        {
+          return true;
+        }
+        // note that slicing a layout, while it may increase _abs_stride,
```
Comment (fbusato) [inline cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: please use 120 columns width

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh [hunk 4]
```diff
@@ -145,26 +148,50 @@ template <typename _Tp, typename _Extents, typename _LayoutPolicy, typename _Acc
     namespace cudax       = ::cuda::experimental;
     const auto __tensor   = cudax::__to_raw_tensor(__mdspan);
     const auto __sorted   = cudax::__sort_by_stride(__tensor);
-    using __stride_t      = decltype(__sorted.__strides[0]);
+    using __stride_t      = ::cuda::std::remove_cvref_t<decltype(__sorted.__strides[0])>;
```
Comment (fbusato) [inline cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: missing header

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh [hunk 5]
```diff
@@ -145,26 +148,50 @@ template <typename _Tp, typename _Extents, typename _LayoutPolicy, typename _Acc
...
       {
-        return true;
+        const auto __abs_stride = cudax::__abs_integer(__strides[__i]);
+        if (__abs_stride <= __max_dist)
+        {
+          return true;
+        }
+        // note that slicing a layout, while it may increase _abs_stride,
+        // cannot increase dimension's contribution to max_dist
+        __max_dist += (extent - 1) * __abs_stride;
```
Comment (fbusato) [inline cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]:
```suggestion
        __max_dist += (__extent - 1) * __abs_stride;
```

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh [hunk 6]
```diff
@@ -145,26 +148,50 @@ template <typename _Tp, typename _Extents, typename _LayoutPolicy, typename _Acc
...
+        return false;
       }
     }
-    for (__rank_t __i = 0; __i + 1 < __rank; ++__i)
+    // max_dist is maximal distance between two elements in sub-layout
+    // restricted to i - 1 dims with smallest strides
+    // max_offset = sum((e_j - 1) * s_j for j in range(i) if s_j > 0)
+    // min_offset = sum((e_j - 1) * s_j for j in range(i) if s_j < 0)
+    // max_dist = max_offset - min_offset = sum((e_j - 1) * |s_j| for j in range(i))
+    __stride_t __max_dist = 0;
```
Comment (fbusato) [inline cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: this should be the extent type because it needs to be representable by `required_span_size()`

**Review by fbusato — COMMENTED** [review]
The improvement makes sense at high-level, but I'm still not 100% sure about the implications of having situations like the example you provided. Namely, a row that "starts" before the previous one ends. It is not overlapping at index level, but still hard to predict potentially issues.

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh [hunk 7]
```diff
@@ -125,14 +127,15 @@ __sort_by_stride(const __raw_tensor<_ExtentT, _StrideT, _Tp, _MaxRank>& __tensor
...
 }
 
-//! @brief Conservative check for interleaved stride order in tensor layouts.
+//! @brief Conservative check for non-unique/interleaved layout.
 //!
-//! Sorts modes by ascending absolute stride, then verifies two conditions:
-//! 1. No mode with extent > 1 has stride == 0 (broadcast)
-//! 2. No mode's span (extent * |stride|) exceeds the next mode's |stride|
+//! A layout is non-unique/interleaved if there exists two different indices
+//! that map to the same element (i.e. `dot(idx1, strides) == dot(idx2, strides)`).
```
Comment (mhoemmen) [inline cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: I don't quite understand the purpose of this check.  Is the goal to check the preconditions of the strides in the `layout_stride::mapping(extents, array)` constructor?  Is the goal to check the strides of a general `is_strided() == true` mapping?  Or is the goal to check whether an arbitrary mapping is unique?

If the goal is to check whether an arbitrary mapping is unique, then why not use the `is_unique()` query?

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh (hunk 6 above)
Comment (mhoemmen) [reply cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: Arbitrary layout mappings aren't necessarily strided, so I'm not sure if speaking of "strides" would generally make sense.

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh (hunk 7 above)
Comment (mhoemmen) [reply cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: I think I understand now.  Is the goal to test specifically `layout_stride_relaxed::mapping` to see whether it's unique?  If so, then this test might make sense.

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh (hunk 6 above)
Comment (mhoemmen) [reply cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: I think I understand now.  Is the goal to test specifically `layout_stride_relaxed::mapping` to see whether it's unique?  If so, then this test might make sense.

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh [hunk 8]
```diff
@@ -125,14 +127,15 @@ __sort_by_stride(const __raw_tensor<_ExtentT, _StrideT, _Tp, _MaxRank>& __tensor
...
 //!
-//! Returns true when the layout fails this non-interleaving rule. This is stronger than a mathematical injectivity
-//! check and may reject some layouts with distinct offsets.
+//! The check is exact for layouts that are dense or are a result of common
+//! tensor layout transformations (transposing, slicing, broadcasting, reshaping,
+//! vectorizing) applied to a dense layout. In general case, it may
+//! incorrectly return true for non-interleaved/unique layouts.
 //!
 //! @param[in] __mdspan Mdspan view to inspect
```
Comment (mhoemmen) [inline cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: If I have an `mdspan` with arbitrary layout, then I can just ask its layout mapping if `is_unique()` is `true`.

Layout mappings can implement that check conservatively by returning `false`, but It's up to the layout mapping to decide that.  For the special case of `layout_stride_relaxed`, the right approach would be for `layout_stride_relaxed::mapping` to have this code.

**Comment by mhoemmen** [comment]
@Jacobfaib wrote:
> @mhoemmen could you take a look at this?

Thanks for asking! : - )

I'm not quite sure what's going on here.

If I have an `mdspan` with arbitrary layout, then I can just ask its layout mapping if `is_unique()` is `true`.  Layout mappings can implement that check conservatively by returning `false`, but It's up to the layout mapping to decide that.

For the special case of `layout_stride_relaxed`, the right approach would be for `layout_stride_relaxed::mapping::is_unique` to implement this check.

For user-defined strided layouts (i.e., layouts that aren't `layout_stride_relaxed`, `layout_stride`, `layout_{left,right}`, or `layout_{left,right}_padded`), again, the right approach would be to trust what the user-defined layout mapping's `is_strided()` function returns.

**Comment by stiepan** [comment]
@mhoemmen thanks for the overview! @fbusato I like the idea to move this (relaxed for step > 1) check to layout_stride_realaxed::is_unique and just use whatever is_unique we have in the mdspan/fill_bytes validation, do you see any reason not to?

**Comment by fbusato** [comment]
> I like the idea to move this (relaxed for step > 1) check to layout_stride_realaxed::is_unique and just use whatever is_unique we have in the mdspan/fill_bytes validation, do you see any reason not to?

I don't think there any issue with that. We can move the code to `is_unique()`

One last note, we should also add the check to `layout_stride::is_unique`, even better a common function

**Comment by mhoemmen** [comment]
@fbusato wrote:
> One last note, we should also add the check to `layout_stride::is_unique`, even better a common function

`layout_stride::mapping::is_unique()` always returns `true`.  The mapping is always unique; this is a precondition on the strides in `layout_stride::mapping`'s constructor.

Please also note the proposed resolution for [LWG 4603](https://cplusplus.github.io/LWG/lwg-active.html#4603), "`layout_stride` should accept zero strides for empty extents."  `layout_stride::mapping` remains always unique even if its extents object has size zero.

**Comment by fbusato** [comment]
thanks @mhoemmen. However, `layout_stride` implements a weaker check for unique mapping, see  https://github.com/NVIDIA/cccl/blob/main/libcudacxx/include/cuda/std/__mdspan/layout_stride.h#L281. For instance, the example provided by @stiepan fails. 
My suggestion is to add a common function shared across `layout_stride` and `layout_stride_relaxed`

**Comment by mhoemmen** [comment]
@fbusato wrote:
> For instance, the example provided by @stiepan fails.

That example violates the preconditions of `layout_stride::mapping`'s constructor, per [[mdspan.layout.stride.cons] 4.3](https://eel.is/c++draft/views.multidim#mdspan.layout.stride.cons-4.3).  `5 >= 2 * 3` is `false`, and `2 >= 5 * 2` is `false`, so there is no permutation of (0, 1) that makes the condition in [mdspan.layout.stride.cons] 4.3 hold.

The intent of `layout_stride` is to be able to represent any layout mapping resulting from zero or more applications of `submdspan` to a `layout_left` or `layout_right` mapping.  @stiepan 's example could never arise from this use case.

That being said, @stiepan 's example shows that [Note 1](https://eel.is/c++draft/views.multidim#mdspan.layout.stride.cons-note-1) on [mdspan.layout.stride.cons] 4.3 is wrong.  We should file an LWG issue to fix the Note.

**Comment by stiepan** [comment]
@mhoemmen 

>The intent of layout_stride is to be able to represent any layout mapping resulting from zero or more applications of submdspan to a layout_left or layout_right mapping. @stiepan 's example could never arise from this use case.

I didn't dig deep into c++ standard, but it doesn't seem obvious to me that it can't. It's a result of slicing with step = 2, so in terms of C++ mdspan it would be something like this, I think:

* Start with mdspan of `std::mdspan<int, std::extents<size_t, 2, 5>, std::layout_right>`
* Take `submdspan` with full_extent, followed by `strided_slice{.offset=0, .extent=5, .stride=2}`
  * New rightmost extent: `1 + (s.extent - 1) / s.stride = 1 + ((5 - 1) / 2) = 3` (as per https://en.cppreference.com/cpp/container/mdspan/strided_slice)
  * New rightmost stride = 2
  * The leftmost stirde is unchanged = 5. We get the `2 * 3 > 5 and 5 * 2 > 2`.
  
  
NB. It feels like the layout_stride::mapping precondition is too restrictive - it breaks for slicing with step > 1 in a way that feels arbitrary. In the above example, if the initial rightmost extent was 6 not 5 - the precondition works. For 5 it breaks. Even though we perform logically the same "slice full extent with stride 2" operation.

**Comment by mhoemmen** [comment]
@stiepan wrote:
> New rightmost extent: 1 + (s.extent - 1) / s.stride = 1 + ((5 - 1) / 2) = 3 (as per https://en.cppreference.com/cpp/container/mdspan/strided_slice)

`strided_slice` doesn't exist!  We replaced it.  See [P3982](https://wg21.link/p3982).  You can find the spec here: https://eel.is/c++draft/views.multidim .

Tomasz Kamiński and his colleagues implemented `extent_slice` in GCC.  You can try it with GCC 16.1: https://godbolt.org/z/5xrMToso7 .

```c++
  using extents_type = std::extents<std::size_t, 2, 5>;
  extents_type exts;
  std::layout_right::mapping<extents_type> map(exts);
  std::vector<float> storage(map.required_span_size());

  std::mdspan x(storage.data(), map);
  assert(x.extent(0) == 2);
  assert(x.extent(1) == 5);
  assert(x.stride(0) == 5);
  assert(x.stride(1) == 1);

  std::extent_slice slice{.offset=0, .extent=5, .stride=2};
  auto x_sub = std::submdspan(x, std::full_extent, slice);
```

This actually violates the preconditions of `submdspan`.  `extent_slice::extent` denotes the _output_ extent.  Thus, the slice represents the indices (0, 2, 4, 6, 8).  It's out of bounds!

**Comment by stiepan** [comment]
@mhoemmen Thanks for the pointers!

So, the extent_slice describes extent in terms of output. Shouldn't this be legal then?

```
  using extents_type = std::extents<std::size_t, 2, 5>;
  extents_type exts;
  std::layout_right::mapping<extents_type> map(exts);
  std::vector<float> storage(map.required_span_size());

  std::mdspan x(storage.data(), map);
  assert(x.extent(0) == 2);
  assert(x.extent(1) == 5);
  assert(x.stride(0) == 5);
  assert(x.stride(1) == 1);

  // out of [0, 1, 2, 3, 4] cut out every second -> [0, 2, 4]
  // it doesn't really go out of bounds
  // in part. 0 + 1 + (3 - 1) * 2 = 5 <= 5 (original extent)
  // as per https://eel.is/c++draft/views.multidim#mdspan.sub.overview-8.1.4
  std::extent_slice slice{.offset=0, .extent=3, .stride=2};
  auto x_sub = std::submdspan(x, std::full_extent, slice);

  // and yet we have "offending" layout 2 * 3 > 5
  assert(x_sub.extent(0) == 2);
  assert(x_sub.extent(1) == 3);
  assert(x_sub.stride(0) == 5);
  assert(x_sub.stride(1) == 2);
  
  assert(x_sub.mapping()(0, 0) == 0);
  assert(x_sub.mapping()(0, 1) == 2);
  assert(x_sub.mapping()(0, 2) == 4);
  assert(x_sub.mapping()(1, 0) == 5);
  assert(x_sub.mapping()(1, 1) == 7);
  assert(x_sub.mapping()(1, 2) == 9);
```

**Comment by fbusato** [comment]
> strided_slice doesn't exist! We replaced it. See [P3982](https://wg21.link/p3982). You can find the spec here: [eel.is/c++draft/views.multidim](https://eel.is/c++draft/views.multidim) .

we don't have it yet😕. Let me create a new issue for it

**Comment by mhoemmen** [comment]
@stiepan wrote:
> Shouldn't this be legal then?

I think you're right!  I stand corrected.  The precondition looks wrong here.

I'm in the middle of discussing the LWG issue on the e-mail list; I'll update the discussion.

Here's a Compiler Explorer link with your example: https://godbolt.org/z/Ts8YGrcx6 .

**Comment by mhoemmen** [comment]
A WG21 colleague suggests changing the test from `s[` $p_i$ `] >= s[` $p_{i-1}$ `] * e.extent(` $p_{i-1}$ `)` to `s[` $p_i$ `] >= 1 + (e.extent(` $p_{i-1}$ `) - 1) * s[` $p_{i-1}$ `]`.

https://godbolt.org/z/1q9s77E14

```
  assert(! (x_sub.stride(1) >= 1 + (x_sub.extent(0) - 1) * x_sub.stride(0)));
  assert(   x_sub.stride(0) >= 1 + (x_sub.extent(1) - 1) * x_sub.stride(1));
```

**Comment by stiepan** [comment]
Thanks @mhoemmen for following-up on this!

In this PR ([L176](https://github.com/NVIDIA/cccl/pull/9788/changes#diff-89fa8ad79f01780b836bc4d99e1b0caff38a307fe0cb0e7561dbbd8fce4e8540R176))  the check is: 

1. `s[` $p_i$ `] >` $\sum_{j=0}^{i - 1}$ `(e.extent(` $p_{j}$ `) - 1) * s[` $p_{j}$ `]`,

vs the check suggested by person from WG21:

2. `s[` $p_i$ `] >= 1 + (e.extent(` $p_{i-1}$ `) - 1) * s[` $p_{i-1}$ `]`.

(changing `>= + 1` to equiv `>`) 
`s[` $p_i$ `] > (e.extent(` $p_{i-1}$ `) - 1) * s[` $p_{i-1}$ `]`

So it's either that we look at the sum of "max offsets" for all previous dims or only at the last term in the sum - the dim next to us. I think we must look at the whole sum, otherwise the check is too weak (accepts non-unique layouts). Consider the following

```
extents = (2, 2, 2)
strides = (3, 2, 1)
```

Note, layout is not unique: (1, 0, 0) -> 1 * 3 = 3 and (0, 1, 1) = 2 + 1 = 3. The check with the sum correctly rejects it
`3 > 1 * 1 + 1 * 2 = 3 is false`. However, the check that looks only at the last extent
`3 >= 1 + 1 * 2 is true`.

As for the check with the full sum, I added a sketch of a proof that it rejects all non-unique layouts in a comment.

**Comment by stiepan** [comment]
A bit off-topic comment:

If we reformulate the `s[` $p_i$ `] >` $\sum_{j=0}^{i - 1}$ `(e.extent(` $p_{j}$ `) - 1) * s[` $p_{j}$ `]`,  condition in terms of `>=` , i.e

`s[` $p_i$ `] >= 1+ ` $\sum_{j=0}^{i - 1}$ `(e.extent(` $p_{j}$ `) - 1) * s[` $p_{j}$ `]` it gives nice symmetric counterpart for `is_exhaustive`
`s[` $p_i$ `] <= 1 + ` $\sum_{j=0}^{i - 1}$ `(e.extent(` $p_{j}$ `) - 1) * s[` $p_{j}$ `]`.

If the layout_stride happens to be a permutation of layout_right/layout_left, both of the conditions hold with equality. But the is_exhaustive formulated this way can correctly report as exhaustive some layouts that are non-unique, while the condition I found in views.multdim
image
enforces uniquness (being permutation of right/left) as well.

**Comment by mhoemmen** [comment]
@stiepan We've very much botched the wording for `layout_stride`, alas.  Thanks for pointing this out!

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh (hunk 1 above)
Comment (stiepan) [reply cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: ooops, leftover, removed!

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh (hunk 2 above)
Comment (stiepan) [reply cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: I went with auto = static_cast pattern but changed it to index_t

File: cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh (hunk 4 above)
Comment (stiepan) [reply cudax/include/cuda/experimental/__copy_bytes/tensor_query.cuh]: Added it.

**Comment by stiepan** [comment]
@fbusato I addressed the comments, but, for now, kept it as a separate helper. 

>My suggestion is to add a common function shared across layout_stride and layout_stride_relaxed

I like the idea to just rely on is_unique (it feels like exactly right tool for that). I see no issue in moving this helper inside `layout_stride_relaxed::mapping::is_unique`. For `layout_stride` though I am not sure how far can I go with adjustments. Currently its constructor checks `__check_unique_mapping` using the too restrtictive condition. As the condition implies uniqueness, is_unique just returns `true`. Where should the less restrictive condition go - just replace the old check in the constructor?

**Comment by stiepan** [comment]
>@stiepan We've very much botched the wording for layout_stride, alas. Thanks for pointing this out!

I am happy to help!

**Comment by fbusato** [comment]
> I like the idea to just rely on is_unique (it feels like exactly right tool for that). I see no issue in moving this helper inside layout_stride_relaxed::mapping::is_unique. For layout_stride though I am not sure how far can I go with adjustments.

I don't see a big problem here. We are relaxing the condition, not making it more restricting. We current implemented `is_unique` as `true` but it is not quite right.

> Currently its constructor checks __check_unique_mapping using the too restrtictive condition. As the condition implies uniqueness, is_unique just returns true. Where should the less restrictive condition go - just replace the old check in the constructor?

my idea would be adding the code to `is_unique()` and replace `__check_unique_mapping` precondition checking with it. If the code is common/very similar between  `layout_stride_relaxed` and `layout_relaxed`, we can create a common function.
Let me know if you have a different idea.

**Comment by mhoemmen** [comment]
@stiepan wrote:
> ... vs the check suggested by person from WG21:

Our WG21 colleague suggested permuting all the 1 extents to the end, because their corresponding strides don't matter.  (We don't care about ANY of the strides if _any_ extent is zero.  For extents of 1, their corresponding strides don't matter, because they would only ever get multiplied by zero.)

I had a different wording suggestion that would have had the same effect.  As usual, Tomasz's wording is more concise : - ) .

Regarding `is_exhaustive`, we have an open LWG issue for this, [LWG 4562](https://cplusplus.github.io/LWG/lwg-active.html#4562).  The author suggests just having `is_exhaustive()` return whether `required_span_size()` equals the size of the multidimensional index space.

**Comment by mhoemmen** [comment]
I wrote:
> Regarding is_exhaustive, we have an open LWG issue for this, [LWG 4562](https://cplusplus.github.io/LWG/lwg-active.html#4562). The author suggests just having is_exhaustive() return whether required_span_size() equals the size of the multidimensional index space.

This only works for `layout_stride`, btw, not for `layout_stride_relaxed`.  Counterexample for the latter: positive offset of N-1, size of N, strides all 0.

**Comment by stiepan** [comment]
@mhoemmen 

Regarding the `is_unique` - agree on the 0 extents (no point in looking at any strides) and 1 extents (corresponding stride is irrrelevant)! But the condition:

image

is too weak I think, even if we ignore extents 0 and 1. In the [example I provided,](https://github.com/NVIDIA/cccl/pull/9788#issuecomment-4975085864) the extents are all 2s.

```
extents = (2, 2, 2)
strides = (3, 2, 1)

# idx: (1, 0, 0) and (0, 1, 1) clash
```

or

```
extents = (3, 3, 3)
strides = (7, 3, 1)

#idx clash: (1, 0, 0) and (0, 2, 1)
```

I think one really needs to look at the sum of all those "max offsets" generated by previous extents, not just the one before.

>Regarding is_exhaustive, we have an open LWG issue for this, [LWG 4562](https://cplusplus.github.io/LWG/lwg-active.html#4562). The author suggests just having is_exhaustive() return whether required_span_size() equals the size of the multidimensional index space.

That's a clever trick to avoid some explicit formulas (since `layout_stride` must be unique anyway)!

File: libcudacxx/include/cuda/std/__mdspan/layout_stride.h [hunk 9]
```diff
@@ -284,16 +284,25 @@ class _CCCL_DECLSPEC_EMPTY_BASES layout_stride::mapping
...
+      {
+        continue;
+      }
+      const auto __stride = static_cast<index_type>(__strides()[__permute[__i]]);
+      if (__stride < __span_size)
       {
         __result = false;
         break;
       }
+      __span_size += (__extent - 1) * __stride;
```
[the remark this answers is not in the document]
Comment (stiepan) [reply libcudacxx/include/cuda/std/__mdspan/layout_stride.h:305]: Great catch, thanks, I've missed that when adjusting the code.

File: libcudacxx/test/libcudacxx/cuda/containers/views/mdspan/layout_stride_relaxed/properties.pass.cpp [hunk 10]
```diff
@@ -95,90 +104,172 @@ TEST_FUNC constexpr bool test()
...
+  test_layout_mapping_stride_relaxed(
+    cuda::std::extents<int, 5, 2, 1>{}, cuda::std::array<intptr_t, 3>{15, 9, -72}, 0, true, true);
+  // right(5, 5, 3)[:, ::3, ::3]
+  test_layout_mapping_stride_relaxed(
+    cuda::std::extents<int, 5, 2, 1>{}, cuda::std::array<intptr_t, 3>{15, 9, 0}, 0, true, true);
+
+  // Overlapping extents (non-unique layouts)
+  // Counterexamples to attempt at checking 1 + (e - 1) * stride <= next_stride
+  // for uniqueness - examples below meet the condition, but are non-unique
+  test_layout_mapping_stride_relaxed(
```
Comment (stiepan) [inline libcudacxx/test/libcudacxx/cuda/containers/views/mdspan/layout_stride_relaxed/properties.pass.cpp:267]: @mhoemmen 

Here are a couple of negative tests targeting the `1 + (e-1)*stride` check (as I found it here https://cplusplus.github.io/LWG/lwg-active.html#4606) - such a check would incorrectly see them as unique.

Comment (stiepan) [reply libcudacxx/test/libcudacxx/cuda/containers/views/mdspan/layout_stride_relaxed/properties.pass.cpp:267]: Update: [LWG 4606](https://cplusplus.github.io/LWG/lwg-active.html#4606) was reworded to spell out "greater than sum of", it looks OK now.

File: libcudacxx/include/cuda/__mdspan/layout_stride_relaxed.h [hunk 11]
```diff
@@ -159,6 +160,104 @@ class _CCCL_DECLSPEC_EMPTY_BASES layout_stride_relaxed::mapping
...
+  //! @param[in] __tensor Raw tensor whose stride order is inspected
+  //! @return Mode permutation sorted by ascending absolute stride
+  [[nodiscard]] _CCCL_HOST_API constexpr ::cuda::std::array<rank_type, __rank_> __get_stride_order() const noexcept
+  {
+    ::cuda::std::array<rank_type, __rank_> __perm{};
+    for (rank_type __i = 0; __i < __rank_; ++__i)
+    {
+      __perm[__i] = __i;
+    }
+    ::cuda::std::sort(__perm.begin(), __perm.begin() + __rank_, [this](rank_type __a, rank_type __b) {
```
Comment (miscco) [inline libcudacxx/include/cuda/__mdspan/layout_stride_relaxed.h:186]: I believe we should just  implement a stupid plain sort, given that the rank is generally a small number that will avoid all the host device issues

**Review by miscco — CHANGES_REQUESTED** [review]

File: libcudacxx/test/libcudacxx/cuda/containers/views/mdspan/layout_stride_relaxed/properties.pass.cpp (hunk 10 above)
Comment (mhoemmen) [reply libcudacxx/test/libcudacxx/cuda/containers/views/mdspan/layout_stride_relaxed/properties.pass.cpp:267]: @stiepan Tomasz fixed it : - )  Thanks for working on this!

File: libcudacxx/include/cuda/__mdspan/layout_stride_relaxed.h (hunk 11 above)
Comment (fbusato) [reply libcudacxx/include/cuda/__mdspan/layout_stride_relaxed.h:186]: the lambda expression is particularly problematic here. Insertion or bubble sort are also fine

File: libcudacxx/include/cuda/__mdspan/layout_stride_relaxed.h [hunk 12]
```diff
@@ -159,6 +160,104 @@ class _CCCL_DECLSPEC_EMPTY_BASES layout_stride_relaxed::mapping
...
+      }
+    }
+    return false;
+  }
+
+  //! @brief Computes the mode permutation that orders a tensor by ascending absolute stride.
+  //!
+  //! @param[in] __tensor Raw tensor whose stride order is inspected
+  //! @return Mode permutation sorted by ascending absolute stride
+  [[nodiscard]] _CCCL_HOST_API constexpr ::cuda::std::array<rank_type, __rank_> __get_stride_order() const noexcept
```
Comment (fbusato) [inline libcudacxx/include/cuda/__mdspan/layout_stride_relaxed.h:179]:
```suggestion
  [[nodiscard]] _CCCL_API constexpr ::cuda::std::array<rank_type, __rank_> __get_stride_order() const noexcept
```

File: libcudacxx/include/cuda/__mdspan/layout_stride_relaxed.h [hunk 13]
```diff
@@ -159,6 +160,104 @@ class _CCCL_DECLSPEC_EMPTY_BASES layout_stride_relaxed::mapping
...
+      const auto __abs_stride_b = ::cuda::uabs(strides().stride(__b));
+      return __abs_stride_a < __abs_stride_b;
+    });
+    return __perm;
+  }
+
+  //! @brief Conservative uniqueness check (i.e. may return false for some unique layouts)
+  //! It returns false for any non-unique layout and correctly reports as unique
+  //! layouts that can be derived as permutations and slices of layout_right/layout_left.
+  [[nodiscard]] _CCCL_HOST_API constexpr bool __is_unique() const noexcept
```
Comment (fbusato) [inline libcudacxx/include/cuda/__mdspan/layout_stride_relaxed.h:197]:
```suggestion
  [[nodiscard]] _CCCL_API constexpr bool __is_unique() const noexcept
```

File: libcudacxx/include/cuda/std/__mdspan/layout_stride.h [hunk 14]
```diff
@@ -284,16 +284,25 @@ class _CCCL_DECLSPEC_EMPTY_BASES layout_stride::mapping
     array<rank_type, extents_type::rank()> __permute{_Pos...};
     __bubble_sort_by_strides(__permute);
```
Comment (fbusato) [inline libcudacxx/include/cuda/std/__mdspan/layout_stride.h:285]: indeed, we already have bubble sort

File: libcudacxx/test/libcudacxx/cuda/containers/views/mdspan/layout_stride_relaxed/properties.pass.cpp [hunk 15]
```diff
@@ -30,8 +30,17 @@ using cuda::std::intptr_t;
...
 template <class E>
 TEST_FUNC constexpr void test_layout_mapping_stride_relaxed(
-  E ext, [[maybe_unused]] cuda::std::array<intptr_t, E::rank()> input_strides, intptr_t offset, bool expected_is_strided)
+  E ext,
+  [[maybe_unused]] cuda::std::array<intptr_t, E::rank()> input_strides,
+  intptr_t offset,
+  bool expected_is_strided,
+  bool expected_is_unique)
 {
+#if defined(__CUDA_ARCH__)
```
Comment (fbusato) [inline libcudacxx/test/libcudacxx/cuda/containers/views/mdspan/layout_stride_relaxed/properties.pass.cpp:39]: update the test after using a simple sort function

**Review by fbusato — CHANGES_REQUESTED** [review]
looks good. I think the only missing piece is the sorting routine compatible with both host/device.
